pallets/assets-transaction-payment/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+		Ok(self.signature_scheme)
+	}
+
+	fn validate(
+		&self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> TransactionValidity {
+		use pallet_transaction_payment::ChargeTransactionPayment;
+		let (fee, inital_payment) = self.withdraw_fee(who, call, info, len)?;
+		// Check if the given asset is valid
+		if let InitialPayment::Asset(asset) = inital_payment {
+			let allowed_assets = <AllowedAssets<T>>::get();
+			if !allowed_assets.contains(&asset.asset()) {
+				return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+			}
+		}
+		let priority = ChargeTransactionPayment::<T>::get_priority(info, len, self.tip, fee);
+		Ok(ValidTransaction { priority, ..Default::default() })
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let (_fee, initial_payment) = self.withdraw_fee(who, call, info, len)?;
+		// Check if the given asset is valid
+		if let InitialPayment::Asset(asset) = &initial_payment {
+			let allowed_assets = <AllowedAssets<T>>::get();
+			if !allowed_assets.contains(&asset.asset()) {
+				return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+			}
+		}
+		Ok((self.tip, who.clone(), initial_payment))
+	}
+
+	fn post_dispatch(
+		pre: Option<Self::Pre>,
+		info: &DispatchInfoOf<Self::Call>,
+		post_info: &PostDispatchInfoOf<Self::Call>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		if let Some((tip, who, initial_payment)) = pre {
+			match initial_payment {
+				InitialPayment::Native(already_withdrawn) => {
+					pallet_transaction_payment::ChargeTransactionPayment::<T>::post_dispatch(
+						Some((tip, who, already_withdrawn)),
+						info,
+						post_info,
+						len,
+						result,
+					)?;
+				},
+				InitialPayment::Asset(already_withdrawn) => {
+					let actual_fee = pallet_transaction_payment::Pallet::<T>::compute_actual_fee(
+						len as u32, info, post_info, tip,
+					);
+					let asset_id = already_withdrawn.asset();
+					T::OnChargeAssetTransaction::correct_and_deposit_fee(
+						&who,
+						info,
+						post_info,
+						actual_fee.into(),
+						tip.into(),
+						already_withdrawn.into(),
+					)?;
+					Pallet::<T>::deposit_event(Event::<T>::AssetTxFeePaid {
+						who,
+						actual_fee,
+						tip,
+						asset_id,
+					});
+				},
+				InitialPayment::Nothing => {
+					// `actual_fee` should be zero here for any signed extrinsic. It would be
+					// non-zero here in case of unsigned extrinsics as they don't pay fees but
+					// `compute_actual_fee` is not aware of them. In both cases it's fine to just
+					// move ahead without adjusting the fee, though, so we do nothing.
+					debug_assert!(tip.is_zero(), "tip should be zero if initial fee was zero.");
```

Comment:
Passed an error back as well as emitted an event

Reply:
Logic here changed and assertion removed at all. Here is a new implementation by @nakul1010:

```rust
	InitialPayment::Nothing => {
	// `actual_fee` should be zero here for any signed extrinsic. It would be
	// non-zero here in case of unsigned extrinsics as they don't pay fees but
	// `compute_actual_fee` is not aware of them. In both cases it's fine to just
	// move ahead without adjusting the fee, though, so we do nothing.
	return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
},

```

Which is really different from the initial implementation from @Gauthamastro and is not match comment description. Is it was discussed and decided to be changed to return error here?